File: A1382*balance-a-binary-search-tree/solution.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution(object):
    def inorder(self, node, lst):
        if node is None:
            return
        
        self.inorder(node.left, lst)
        lst.append(node.val)
        self.inorder(node.right, lst)

    def _list_to_tree(self, lst, low, high):
        if low > high:
            return

        mid = (high + low) / 2
        root = TreeNode(lst[mid])
        root.left = self._list_to_tree(lst, low, mid - 1)
        root.right = self._list_to_tree(lst, mid + 1, high)
        return root
    
    def ordered_list_to_balanced_tree(self, lst):
        l, h = 0, len(lst) - 1
        mid = (h + l) / 2
        root = TreeNode(lst[mid])
        root.left = self._list_to_tree(lst, l, mid - 1)
        root.right = self._list_to_tree(lst, mid + 1, h)
        return root
    
    def balanceBST(self, root):
        """
        :type root: TreeNode
        :rtype: TreeNode
        """
        
        lst = []
        self.inorder(root, lst)
        #print(lst)        
        
        ret = self.ordered_list_to_balanced_tree(lst)
        return ret

File: A1382*balance-a-binary-search-tree/solution.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public void inorder(TreeNode node, ArrayList lst) {
        if (node == null) {
            return;
        }
        inorder(node.left, lst);
        lst.add(node.val);
        inorder(node.right, lst);
    }

    public TreeNode _lstToBST(ArrayList<Integer> lst, int low, int high) {
        if (low > high) {
            return null;
        }
        
        int mid = (low + high) / 2;
        //System.out.println(lst.get(mid));
        TreeNode root = new TreeNode(lst.get(mid));
        root.left = _lstToBST(lst, low, mid - 1);
        root.right = _lstToBST(lst, mid + 1, high);
        return root;    
    }
    
    public TreeNode orderedListToBalanceBST(ArrayList<Integer> lst) {
        int low = 0;
        int high = lst.size() - 1;
        int mid = (low + high) / 2;
        
        TreeNode root = new TreeNode(lst.get(mid));
        root.left = _lstToBST(lst, low, mid - 1);
        root.right = _lstToBST(lst, mid + 1, high);
        return root;
    }
    
    public TreeNode balanceBST(TreeNode root) {
        ArrayList<Integer> lst = new ArrayList<Integer>();
        inorder(root, lst);
        
        TreeNode ret = orderedListToBalanceBST(lst);
        return ret;
    }
}
